### DISCUSSION TITLE
SCIM 2.0 Support as client and server

### DISCUSSION THREAD

**Discussion by fforootd** [zitadel/zitadel#1931] [Ideas]
At some point we should provide a SCIM 2.0 client and server Addon

We could reuse existing projects like https://github.com/elimity-com/scim

**Comment by fforootd** [comment]
Yeah, we did not yet deeply inspect the project but it could be that we integrate something like this in the future.

**Comment by hifabienne** [comment]
One customer is interested in a SCIM API. They would just need the user information in their application.
Preferable ZITADEL would call their app, but it would also be possible to call ZITADEL.

One alternative from my perspective would be to implement actions on user create and user update. This would allow the customer to implement the SCIM request in javascript and also be able to add any extensions they need.

**Reply by mffap** [reply]
https://github.com/zitadel/zitadel/issues/6601
Describes this pattern.

**Comment by MatthewAry** [comment]
Please see https://learn.microsoft.com/en-us/azure/active-directory/fundamentals/sync-scim
Having SCIM synchronization would be fabulous!

**Reply by fforootd** [reply]
Hi @MatthewAry what would help you?

ZITADEL have a SCIM API, or ZITADEL writing to a SCIM API?

**Reply by MatthewAry** [reply]
My use case would benefit from an API right now. That way my SaaS can better support enterprise customers and they (the customers) can provision access to our services through their own identity provider.

**Reply by fforootd** [reply]
Ok so it's zitadel having a SCIM api that they can provision their users into it, right?

**Reply by MatthewAry** [reply]
Yes.

**Reply by fforootd** [reply]
Got it, thanks

**Comment by sebapoole** [comment]
Has there been any progress on this? Would be a fantastic addition.

**Reply by fforootd** [reply]
Not so far, the need was to small.

Would you mind sharing with what systems you wanted to use SCIM?

**Comment by adamphetamine** [comment]
Apple just announced that they will support OIDC for creating Managed AppleIDs (MAID) but this will also require SCIM and 'OpenID Shared Signals Framework'

See 17:30 in this video-
https://developer.apple.com/videos/play/wwdc2023/10254

**Reply by fforootd** [reply]
Interesting 

I was not yet aware 😅 thanks for bringing this up.

**Comment by adamphetamine** [comment]
I should briefly add why this is important-
Apple has built an MDM system for business and schools that allows them to upload a serial to your account as soon as a device ships, and while the device is still in the air, you can configure it.
So when it hits a users desk and they connect it into a network, it will download all the necessary profiles, configs, apps etc. without being touched by IT. 
There's a lot of moving parts in the backend, and until this month the only way to federate some of these features was via Azure AD or Google. 

Adding OIDC is a **huge** deal, but I'm not technical enough to know if the requirements for SCIM and the Shared Signals Framework make it difficult to support- I hope not!

**Comment by braginini** [comment]
We at [NetBird](https://github.com/netbirdio/netbird)  would love to have support for SCIM :+1:

**Reply by fforootd** [reply]
Just to be sure, I guess you would love to see a SCIM server in ZITADEL, correct?

**Reply by braginini** [reply]
Both. Right now we are building a group and user sync between IDPs and our management system. So pretty much an internal SCIM server. We would like to sync users and groups from Zitadel.

Later on we could use Zitadel as a bridge between all IDP and then a zitadel SCIM server will come into play.

Does this make sense?

**Comment by mloepfe** [comment]
I also would love to see SCIM (client) support to provision users and groups to Atlassian (Access, Confluence, Jira)

**Comment by K2CanDo** [comment]
For us, SCIM is a strict requirement for the adoption of Zitadel for our upcoming products.

**Reply by fforootd** [reply]
Would you need zitadel as SCIM Server or SCIM Client?

**Reply by K2CanDo** [reply]
Preferably both of course, but for our current set of applications to integrate, a SCIM client would be sufficient atm.

**Comment by HungryHowies** [comment]
+1 on SCIM client.  We  have opensource applications that required user provsioning specially group attributes.

**Reply by compgeniuses** [reply]
+1 SCIM client and Server

**Comment by chicco785** [comment]
We have a cloud-edge solution that in some scenarios is not always connected. Having SCIM client and server support, would allow to synchronise the system (using both Zitadel) while connectivity is available.

**Comment by compgeniuses** [comment]
Looking that almost all solutions, for the Enteprise, as well as new even opensource projects are supporting SCIM, i think it would be in order to have soe priorities setup to have an implementation of the same on Zitadel.

Considering that Zitadel is fronting as a Identity infrastructure.

**Reply by hifabienne** [reply]
We do have the outbound provisioning with SCIM on our roadmap for this year.
https://github.com/zitadel/zitadel/issues/6601
https://github.com/orgs/zitadel/projects/6/views/1

**Reply by compgeniuses** [reply]
ok noted, its in April, too far, but hey, too soon as well.. depends.

**Comment by netsplit** [comment]
A solution where I can provision users from Entra ID into Zitadel would be fabulous.

**Reply by nikzen** [reply]
I would also opt-in here as well. Having a scim /user API per organization would unblock a lot on our side.

**Reply by rolnikrolnik** [reply]
bump, would love that functionality

**Reply by sebapoole** [reply]
Bump as well.

**Reply by HungryHowies** [reply]
Bump same

**Reply by fforootd** [reply]
@hifabienne looking at this thread as well as having talked to some customers we should discuss if priority of https://github.com/orgs/zitadel/projects/6/views/1?pane=issue&itemId=48899908 is still correct or if we want to favor "inbound scim"

WDYT?

**Reply by hifabienne** [reply]
my gut feeling also tells me that inbound scim is more interessting, but looking at the two issues outbound has more votes, but I am not sure if it is only because, we created that issue later than the other. 
Outbound: https://github.com/zitadel/zitadel/issues/6601
Inbound: https://github.com/zitadel/zitadel/issues/8140

**Reply by fforootd** [reply]
Hm good question, I can only share my experience here, and I only heard of inbound scim.

I think outbound scim might become relevant if we start exploring into becoming a b2e tool as well 😁

**Reply by nikzen** [reply]
We also require inbound provisioning. As there is no support for SCIM, we started with a small service to sync data from LDAP to Zitadel (Please be aware that it includes some internal logic on our side, as we also want to directly add a user grant)

The tool can be found here: https://github.com/famedly/ldap-sync

**Comment by nicoonoclaste** [comment]
Not quite a provisionning usecase, but #7909 (managing UNIX users with the IdP) could be solved by a generic helper using OIDC & SCIM:
the latter is required to implement an [NSS module], as OIDC itself does not provide a way to query about the set of users (necessary to implement `getpw{nam, uid}`, as well as `getpwent` which returns the entire set of users)

I'd likely write the necessary helpers, once Zitadel exposes a SCIM API; all that usecase would need is
- `GET` on `/Users/{id}` to get a user's metadata (incl. the groups they belong to) ;
- `GET` on `/Users`, for [enumerating] all users, and [finding] the user with a given name or UID ;
- similarly, support listing all groups, selecting them by name or GID, and listing all members of a given group.

(SCIM does not define a standard schema for numeric UIDs and GIDs, but it's common to define them locally as custom attributes)

[enumerating]: https://scim.dev/playground/users#list-users
[finding]: https://scim.dev/playground/users.html#find-user
[NSS module]: https://www.gnu.org/software/libc/manual/html_node/Name-Service-Switch.html

**Comment by hyperknot** [comment]
I've posted this in the inbound SCIM, but I think it's a better place.

Can someone explain inbound vs. outbound SCIM and server vs. client?

I'm writing from the point of view of a SaaS developer. The following docs on WorkOS seem quite clear about that this use case needs a SCIM server / inbound direction.
[doc1](https://workos.com/blog/the-developers-guide-to-directory-sync-scim)
[doc2](https://workos.com/blog/scim-server)


As I understand the inbound use case means that ZITADEL acts as a SCIM server and IdP platforms act as SCIM clients. ZITADEL provides the REST endpoints and the IdP admins / HR click on their UI for adding/removing users to a given app. As a SaaS developer, inbound direction / SCIM server is what I need. Is that correct?

I mean I really don't understand the outbound direction. Why would anyone need it? Do people really use ZITADEL as a full IdP for their full company? I'm very surprised that there are equal number of votes for inbount/outbound and I suspect that most users would want ZITADEL to serve a SaaS and not manage a full company directory. I mean WorkOS only offers this direction and it seems they are making a killing with onboarding Enterprise clients like Vercel.

TLDR: I'm surprised SCIM server/client is a question here and I'm quite sure ZITADEL could onboard many Enterprise SaaS clients if a built-in SCIM server was offered.

**Reply by hifabienne** [reply]
Hi @hyperknot 
I already answered in the other issue, let me know if you still have questions.

**Reply by hyperknot** [reply]
Thanks a lot for the answer! I don't really see the use case of outbound though, isn't inbound what 99% of users want?

**Reply by hifabienne** [reply]
I guess it highly depends on what you use as you main system, and if you want to be able to manage users in all different systems and keep them in sync with others.

**Reply by hyperknot** [reply]
I understand. I'm just trying to give feedback, as I see that both directions were considered and I believe the example of WorkOS shows that there is a lot of need / money to be made in inbound SCIM.

https://github.com/boxyhq/jackson is also an inbound SCIM.

**Reply by fforootd** [reply]
Thank you for the feedback, I actually agree that for most people in the "CIAM" space "Inbound" SCIM might be more important than "outbound" SCIM. When talking with customers I often encounter that they want to integrate Business Partners and their workforce identity systems with federation for the authentication and SCIM for the provisioning.

**Comment by yann-soubeyrand** [comment]
Hello, we’re investigating ZITADEL as our SSO solution and we would be interested in SCIM, both client and server:
– SCIM server, so that we can provision our users and the users of our customers automatically from Google Workspace, or Microsoft Active Directory, etc.
– SCIM client, so that we can provision users from ZITADEL to AWS identity center, for example.

**Reply by fforootd** [reply]
Hey, thanks for sharing your needs.

Which part is more important to you having the server or client?

**Reply by yann-soubeyrand** [reply]
It’s hard to say, but I guess the client part will be more important to us: we may have several locations where we will want to provision users (AWS, Azure, etc), whereas we can manage user provisioning in ZITADEL with user auto creation configured on the identity providers (we miss user de-provisioning though).

**Reply by fforootd** [reply]
Ok, thank you for this detail. This helps us shape the roadmap

**Comment by f5-ericmarden** [comment]
Another user interested in SCIM. Main use case would be to use Zitadel as the primary IdP, providing SSO and user federation to other services via SAML/SCIM, namely IAM Identity Center. 

SSO with SAML, using Zitadel as the IdP, is already working well for us, but not all SAML service providers will make use of the user provisioning features of the protocol and only do SSO -- leaving us to provision users in the service provider (manually). When this is true, we've found that SCIM support would bridge the gap in some instances.

**Reply by ekarlso** [reply]
Hi, I did a bridge implementation using the aforementioned library https://github.com/elimity-com/scim. If you want I can probably get it shared during the week.

**Reply by compgeniuses** [reply]
why not share what youve done, on this thread so others could take a look at it?

**Comment by hifabienne** [comment]
fyi: We have started the implementation of the following issue: https://github.com/zitadel/zitadel/issues/8140
more endpoints will be added over the next weeks. as well as documenation.
If you have any feedback please let us know.

**Reply by MarkPare** [reply]
This is great! Any estimate on when this will be production-ready?

**Reply by hifabienne** [reply]
SCIM 2.0 Server for user resource, will go from preview to ga with the next major release, e.g by the beginning of july we will have the rc ready, around a month later proper release

**Reply by YaroCabbage** [reply]
On the website it says that this plugin will be available under a commercial license. What does this mean? Will it still be possible to use the plugin with self-hosted Zitadel for free (for my own application)?

**Reply by pal** [reply]
Would love more clarity on the license setup here @hifabienne, see also https://github.com/zitadel/zitadel/issues/9164#issuecomment-4069357802

**Comment by kinthaiofficial** [comment]
SCIM 2.0 support in ZITADEL would be a significant enterprise unlock — it's the standard protocol for provisioning users from IdPs like Okta, Azure AD, and OneLogin.

The main use cases:

**As a SCIM server** (receiving provisioning from an IdP):
- Enterprise customer uses Okta, wants to provision/deprovision ZITADEL users automatically
- HR system manages employees — when someone joins/leaves, ZITADEL reflects that automatically
- SCIM server endpoint: `https://zitadel.example.com/scim/v2/`

**As a SCIM client** (pushing provisioning to downstream apps):
- ZITADEL is the IdP, apps using SCIM get user sync
- App developer doesn't need to implement SAML/OIDC, just SCIM

**SCIM 2.0 endpoints you'd need**:
```
POST   /scim/v2/Users          # Create user
GET    /scim/v2/Users          # List users
GET    /scim/v2/Users/{id}     # Get user
PUT    /scim/v2/Users/{id}     # Replace user
PATCH  /scim/v2/Users/{id}     # Update user (partial)
DELETE /scim/v2/Users/{id}     # Delete user

GET    /scim/v2/Groups         # List groups
POST   /scim/v2/Groups         # Create group
PATCH  /scim/v2/Groups/{id}    # Add/remove members
```

**Authentication for SCIM endpoints**:
Usually Bearer token (a long-lived service account token). ZITADEL could reuse its existing service account mechanism for this.

**Complexity areas**:
- Group ↔ ZITADEL Organization/Project Role mapping isn't 1:1
- SCIM schema extensions for ZITADEL-specific attributes
- Conflict resolution when SCIM and ZITADEL admin both modify a user

Is there an existing tracking issue for this? The LDAP server + SCIM combination would cover most enterprise IdP integration scenarios.
